File: backend/maintenance/resolver.py
#!/usr/bin/env python3
"""
Script Resolver - Vibe Social Network
=====================================
Este script resolve todos os problemas relacionados a:
- Cooldown excessivo de email
- Verificação de email não obrigatória no login
- Sistema de conta ativa
- Limpeza de registros problemáticos
"""

import mysql.connector
import os
import sys
from datetime import datetime, timedelta
import json

class VibeResolver:
    def __init__(self):
        """Inicializa o resolver com configurações do banco"""
        self.db_config = {
            'host': '127.0.0.1',
            'port': 3306,
            'user': 'root',
            'password': 'Dashwoodi@1995',
            'database': 'vibe',
            'charset': 'utf8mb4'
        }
        self.connection = None
        
    def connect_database(self):
        """Conecta ao banco de dados"""
        try:
            print("🔌 Conectando ao banco de dados...")
            self.connection = mysql.connector.connect(**self.db_config)
            print("✅ Conexão estabelecida com sucesso!")
            return True
        except mysql.connector.Error as err:
            print(f"❌ Erro ao conectar ao banco: {err}")
            return False
    
    def execute_query(self, query, params=None, fetch=False):
        """Executa uma query no banco de dados"""
        try:
            cursor = self.connection.cursor(dictionary=True)
            cursor.execute(query, params or ())
            
            if fetch:
                result = cursor.fetchall()
                cursor.close()
                return result
            else:
                self.connection.commit()
                affected_rows = cursor.rowcount
                cursor.close()
                return affected_rows
                
        except mysql.connector.Error as err:
            print(f"❌ Erro na query: {err}")
            return None
    
    def fix_cooldown_issues(self):
        """Remove registros problemáticos que causam cooldown excessivo"""
        print("\n🔧 CORRIGINDO PROBLEMAS DE COOLDOWN")
        print("=" * 50)
        
        # 1. Verificar registros atuais
        records = self.execute_query(
            "SELECT user_id, COUNT(*) as count, MAX(created_at) as last_attempt FROM email_verifications GROUP BY user_id",
            fetch=True
        )
        
        if records:
            print(f"📊 Encontrados registros para {len(records)} usuários:")
            for record in records:
                print(f"   - Usuário {record['user_id']}: {record['count']} tentativas, última: {record['last_attempt']}")
        
        # 2. Remover registros antigos (mais de 1 hora)
        one_hour_ago = datetime.now() - timedelta(hours=1)
        deleted = self.execute_query(
            "DELETE FROM email_verifications WHERE created_at < %s",
            (one_hour_ago,)
        )
        print(f"🗑️ Removidos {deleted} registros antigos (mais de 1 hora)")
        
        # 3. Resetar cooldown para usuários específicos com problemas
        problematic_users = self.execute_query(
            """SELECT DISTINCT user_id FROM email_verifications 
               WHERE created_at > DATE_SUB(NOW(), INTERVAL 1 HOUR) 
               AND user_id IN (25, 26, 27)""",
            fetch=True
        )
        
        if problematic_users:
            user_ids = [str(u['user_id']) for u in problematic_users]
            print(f"🔄 Resetando cooldown para usuários: {', '.join(user_ids)}")
            
            deleted_problem = self.execute_query(
                f"DELETE FROM email_verifications WHERE user_id IN ({', '.join(user_ids)})"
            )
            print(f"✅ Removidos {deleted_problem} registros problemáticos")
        
        print("✅ Problemas de cooldown corrigidos!")
    
    def fix_user_verification_system(self):
        """Corrige o sistema de verificação de usuários"""
        print("\n🔐 CORRIGINDO SISTEMA DE VERIFICAÇÃO")
        print("=" * 50)
        
        # 1. Verificar estrutura da tabela users
        columns = self.execute_query(
            "SHOW COLUMNS FROM users LIKE 'is_verified'",
            fetch=True
        )
        
        if not columns:
            print("📝 Adicionando coluna is_verified na tabela users...")
            self.execute_query(
                "ALTER TABLE users ADD COLUMN is_verified BOOLEAN DEFAULT FALSE"
            )
            print("✅ Coluna is_verified adicionada!")
        else:
            print("✅ Coluna is_verified já existe")
        
        # 2. Adicionar coluna account_status se não existir
        status_columns = self.execute_query(
            "SHOW COLUMNS FROM users LIKE 'account_status'",
            fetch=True
        )
        
        if not status_columns:
            print("📝 Adicionando coluna account_status na tabela users...")
            self.execute_query(
                """ALTER TABLE users ADD COLUMN account_status 
                   ENUM('pending', 'active', 'suspended', 'banned') DEFAULT 'pending'"""
            )
            print("✅ Coluna account_status adicionada!")
        else:
            print("✅ Coluna account_status já existe")
        
        # 3. Atualizar usuários não verificados
        unverified = self.execute_query(
            "SELECT COUNT(*) as count FROM users WHERE is_verified = FALSE OR is_verified IS NULL",
            fetch=True
        )
        
        if unverified and unverified[0]['count'] > 0:
            print(f"📊 Encontrados {unverified[0]['count']} usuários não verificados")
            
            # Marcar contas não verificadas como 'pending'
            updated = self.execute_query(
                """UPDATE users SET account_status = 'pending' 
                   WHERE (is_verified = FALSE OR is_verified IS NULL) 
                   AND account_status != 'suspended' 
                   AND account_status != 'banned'"""
            )
            print(f"🔄 {updated} contas marcadas como 'pending'")
        
        # 4. Ativar contas já verificadas
        verified_inactive = self.execute_query(
            """UPDATE users SET account_status = 'active' 
               WHERE is_verified = TRUE AND account_status = 'pending'"""
        )
        print(f"✅ {verified_inactive} contas verificadas ativadas")
        
        print("✅ Sistema de verificação corrigido!")
    
    def create_account_verification_procedures(self):
        """Cria procedures para verificação de conta"""
        print("\n⚙️ CRIANDO PROCEDURES DE VERIFICAÇÃO")
        print("=" * 50)
        
        # Procedure para verificar se usuário pode fazer login
        can_login_procedure = """
        CREATE OR REPLACE FUNCTION can_user_login(user_id INT) 
        RETURNS JSON
        READS SQL DATA
        DETERMINISTIC
        BEGIN
            DECLARE result JSON;
            DECLARE is_verified BOOLEAN DEFAULT FALSE;
            DECLARE account_status VARCHAR(20) DEFAULT 'pending';
            
            SELECT u.is_verified, u.account_status 
            INTO is_verified, account_status
            FROM users u WHERE u.id = user_id;
            
            IF is_verified = TRUE AND account_status = 'active' THEN
                SET result = JSON_OBJECT(
                    'can_login', TRUE,
                    'verified', TRUE,
                    'status', account_status,
                    'message', 'Login permitido'
                );
            ELSEIF is_verified = FALSE THEN
                SET result = JSON_OBJECT(
                    'can_login', FALSE,
                    'verified', FALSE,
                    'status', account_status,
                    'message', 'Email não verificado'
                );
            ELSEIF account_status != 'active' THEN
                SET result = JSON_OBJECT(
                    'can_login', FALSE,
                    'verified', is_verified,
                    'status', account_status,
                    'message', 'Conta não está ativa'
                );
            ELSE
                SET result = JSON_OBJECT(
                    'can_login', FALSE,
                    'verified', is_verified,
                    'status', account_status,
                    'message', 'Login não permitido'
                );
            END IF;
            
            RETURN result;
        END
        """
        
        try:
            self.execute_query(can_login_procedure)
            print("✅ Function can_user_login criada!")
        except Exception as e:
            print(f"⚠️ Erro ao criar function: {e}")
            print("💡 Continuando sem a function...")
        
        print("✅ Procedures de verificação configuradas!")
    
    def fix_email_service_config(self):
        """Corrige configurações do serviço de email"""
        print("\n📧 CORRIGINDO CONFIGURAÇÕES DE EMAIL")
        print("=" * 50)
        
        # Caminho do arquivo .env
        env_path = "email-service/.env"
        
        # Configuração correta
        correct_config = """# Configurações do banco de dados MySQL
DB_HOST=127.0.0.1
DB_PORT=3306
DB_USER=root
DB_PASSWORD=Dashwoodi@1995
DB_NAME=vibe

# Configurações SMTP
SMTP_HOST=smtp.hostinger.com
SMTP_PORT=587
SMTP_USER=suporte@meuvibe.com
SMTP_PASS=Dashwoodi@1995
SMTP_FROM=no-reply@meuvibe.com

# Configurações de verificação (valores corretos)
VERIFICATION_CODE_EXPIRY=300000
RESEND_COOLDOWN=60000
MAX_RESEND_ATTEMPTS=5

# Porta do serviço
PORT=3001
"""
        
        try:
            with open(env_path, 'w') as f:
                f.write(correct_config)
            print(f"✅ Arquivo {env_path} atualizado com configurações corretas!")
        except Exception as e:
            print(f"⚠️ Erro ao atualizar .env: {e}")
            print("💡 Configure manualmente o arquivo .env do email-service")
        
        print("✅ Configurações de email corrigidas!")
    
    def create_summary_report(self):
        """Cria relatório resumo do estado atual"""
        print("\n📊 RELATÓRIO DE ESTADO ATUAL")
        print("=" * 50)
        
        # Usuários por status
        user_stats = self.execute_query(
            """SELECT 
                account_status,
                is_verified,
                COUNT(*) as count
               FROM users 
               GROUP BY account_status, is_verified
               ORDER BY account_status, is_verified""",
            fetch=True
        )
        
        if user_stats:
            print("👥 Usuários por status:")
            for stat in user_stats:
                verified_text = "✅ Verificado" if stat['is_verified'] else "❌ Não verificado"
                print(f"   - {stat['account_status'].title()}: {stat['count']} usuários ({verified_text})")
        
        # Registros de verificação pendentes
        pending_verifications = self.execute_query(
            "SELECT COUNT(*) as count FROM email_verifications WHERE verified = FALSE",
            fetch=True
        )
        
        if pending_verifications:
            count = pending_verifications[0]['count']
            print(f"📧 Verificações pendentes: {count}")
        
        # Últimas tentativas de verificação
        recent_attempts = self.execute_query(
            """SELECT user_id, email, created_at 
               FROM email_verifications 
               ORDER BY created_at DESC 
               LIMIT 5""",
            fetch=True
        )
        
        if recent_attempts:
            print("📋 Últimas tentativas de verificação:")
            for attempt in recent_attempts:
                print(f"   - Usuário {attempt['user_id']}: {attempt['email']} em {attempt['created_at']}")
        
        print("✅ Relatório concluído!")
    
    def run_all_fixes(self):
        """Executa todas as correções"""
        print("🚀 INICIANDO RESOLUÇÃO DE PROBLEMAS - VIBE")
        print("=" * 60)
        print(f"⏰ Início: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        
        if not self.connect_database():
            return False
        
        try:
            self.fix_cooldown_issues()
            self.fix_user_verification_system()
            self.create_account_verification_procedures()
            self.fix_email_service_config()
            self.create_summary_report()
            
            print("\n" + "=" * 60)
            print("🎉 TODAS AS CORREÇÕES APLICADAS COM SUCESSO!")
            print("=" * 60)
            print("\n✅ Próximos passos:")
            print("1. Reiniciar o serviço de email: cd email-service && npm run dev")
            print("2. Reiniciar o backend principal")
            print("3. Testar login com conta não verificada (deve bloquear)")
            print("4. Testar verificação de email (cooldown de 60s)")
            print("\n💡 Agora o sistema:")
            print("- ✅ Bloqueia login sem verificação de email")
            print("- ✅ Usa cooldown de 60 segundos")
            print("- ✅ Gerencia status da conta corretamente")
            print("- ✅ Remove registros problemáticos automaticamente")
            
            return True
            
        except Exception as e:
            print(f"\n❌ Erro durante a execução: {e}")
            return False
        finally:
            if self.connection:
                self.connection.close()
                print("\n🔌 Conexão com banco fechada")

def main():
    """Função principal"""
    resolver = VibeResolver()
    success = resolver.run_all_fixes()
    
    if success:
        print("\n🎯 Execução concluída com sucesso!")
        sys.exit(0)
    else:
        print("\n💥 Execução falhou. Verifique os logs acima.")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: backend/maintenance/resolver.js
import mysql from 'mysql2/promise';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Script Resolver - Vibe Social Network (Node.js)
 * ===============================================
 * Este script resolve todos os problemas relacionados a:
 * - Cooldown excessivo de email
 * - Verificação de email não obrigatória no login
 * - Sistema de conta ativa
 * - Limpeza de registros problemáticos
 */

class VibeResolver {
    constructor() {
        this.dbConfig = {
            host: '127.0.0.1',
            port: 3306,
            user: 'root',
            password: 'Dashwoodi@1995',
            database: 'vibe',
            charset: 'utf8mb4'
        };
        this.connection = null;
    }

    async connectDatabase() {
        try {
            console.log('🔌 Conectando ao banco de dados...');
            this.connection = await mysql.createConnection(this.dbConfig);
            console.log('✅ Conexão estabelecida com sucesso!');
            return true;
        } catch (err) {
            console.error(`❌ Erro ao conectar ao banco: ${err.message}`);
            return false;
        }
    }

    async executeQuery(query, params = []) {
        try {
            const [result] = await this.connection.execute(query, params);
            return result;
        } catch (err) {
            console.error(`❌ Erro na query: ${err.message}`);
            throw err;
        }
    }

    async fixCooldownIssues() {
        console.log('\n🔧 CORRIGINDO PROBLEMAS DE COOLDOWN');
        console.log('='.repeat(50));

        // 1. Verificar registros atuais
        const records = await this.executeQuery(`
            SELECT user_id, COUNT(*) as count, MAX(created_at) as last_attempt 
            FROM email_verifications 
            GROUP BY user_id
        `);

        if (records.length > 0) {
            console.log(`📊 Encontrados registros para ${records.length} usuários:`);
            records.forEach(record => {
                console.log(`   - Usuário ${record.user_id}: ${record.count} tentativas, última: ${record.last_attempt}`);
            });
        }

        // 2. Remover registros antigos (mais de 1 hora)
        const deleteOldResult = await this.executeQuery(`
            DELETE FROM email_verifications 
            WHERE created_at < DATE_SUB(NOW(), INTERVAL 1 HOUR)
        `);
        console.log(`🗑️ Removidos ${deleteOldResult.affectedRows} registros antigos (mais de 1 hora)`);

        // 3. Resetar cooldown para usuários problemáticos
        const problematicUsers = await this.executeQuery(`
            SELECT DISTINCT user_id 
            FROM email_verifications 
            WHERE created_at > DATE_SUB(NOW(), INTERVAL 1 HOUR) 
            AND user_id IN (25, 26, 27)
        `);

        if (problematicUsers.length > 0) {
            const userIds = problematicUsers.map(u => u.user_id);
            console.log(`🔄 Resetando cooldown para usuários: ${userIds.join(', ')}`);

            const deleteProblemResult = await this.executeQuery(`
                DELETE FROM email_verifications 
                WHERE user_id IN (${userIds.join(',')})
            `);
            console.log(`✅ Removidos ${deleteProblemResult.affectedRows} registros problemáticos`);
        }

        console.log('✅ Problemas de cooldown corrigidos!');
    }

    async fixUserVerificationSystem() {
        console.log('\n🔐 CORRIGINDO SISTEMA DE VERIFICAÇÃO');
        console.log('='.repeat(50));

        // 1. Verificar se a coluna is_verified existe
        const isVerifiedColumns = await this.executeQuery(`
            SHOW COLUMNS FROM users LIKE 'is_verified'
        `);

        if (isVerifiedColumns.length === 0) {
            console.log('➕ Adicionando coluna is_verified...');
            await this.executeQuery(`
                ALTER TABLE users 
                ADD COLUMN is_verified BOOLEAN DEFAULT FALSE
            `);
            console.log('✅ Coluna is_verified adicionada!');
        } else {
            console.log('✅ Coluna is_verified já existe');
        }

        // 2. Verificar se a coluna account_status existe
        const accountStatusColumns = await this.executeQuery(`
            SHOW COLUMNS FROM users LIKE 'account_status'
        `);

        if (accountStatusColumns.length === 0) {
            console.log('➕ Adicionando coluna account_status...');
            await this.executeQuery(`
                ALTER TABLE users 
                ADD COLUMN account_status ENUM('pending', 'active', 'suspended', 'banned') 
                DEFAULT 'pending'
            `);
            console.log('✅ Coluna account_status adicionada!');
        } else {
            console.log('✅ Coluna account_status já existe');
        }

        // 3. Atualizar usuários não verificados
        const unverifiedResult = await this.executeQuery(`
            UPDATE users 
            SET account_status = 'pending' 
            WHERE (is_verified = FALSE OR is_verified IS NULL) 
            AND account_status NOT IN ('suspended', 'banned')
        `);
        console.log(`🔄 ${unverifiedResult.affectedRows} contas marcadas como 'pending'`);

        // 4. Ativar contas já verificadas
        const activatedResult = await this.executeQuery(`
            UPDATE users 
            SET account_status = 'active' 
            WHERE is_verified = TRUE AND account_status = 'pending'
        `);
        console.log(`✅ ${activatedResult.affectedRows} contas verificadas ativadas`);

        console.log('✅ Sistema de verificação corrigido!');
    }

    async fixEmailServiceConfig() {
        console.log('\n📧 CORRIGINDO CONFIGURAÇÕES DE EMAIL');
        console.log('='.repeat(50));

        const envPath = path.join(__dirname, 'email-service', '.env');
        
        const correctConfig = `# Configurações do banco de dados MySQL
DB_HOST=127.0.0.1
DB_PORT=3306
DB_USER=root
DB_PASSWORD=Dashwoodi@1995
DB_NAME=vibe

# Configurações SMTP
SMTP_HOST=smtp.hostinger.com
SMTP_PORT=587
SMTP_USER=suporte@meuvibe.com
SMTP_PASS=Dashwoodi@1995
SMTP_FROM=no-reply@meuvibe.com

# Configurações de verificação (valores corretos)
VERIFICATION_CODE_EXPIRY=300000
RESEND_COOLDOWN=60000
MAX_RESEND_ATTEMPTS=5

# Porta do serviço
PORT=3001
`;

        try {
            fs.writeFileSync(envPath, correctConfig);
            console.log(`✅ Arquivo ${envPath} atualizado com configurações corretas!`);
        } catch (error) {
            console.log(`⚠️ Erro ao atualizar .env: ${error.message}`);
            console.log('💡 Configure manualmente o arquivo .env do email-service');
        }

        console.log('✅ Configurações de email corrigidas!');
    }

    async createSummaryReport() {
        console.log('\n📊 RELATÓRIO DE ESTADO ATUAL');
        console.log('='.repeat(50));

        // Usuários por status
        const userStats = await this.executeQuery(`
            SELECT 
                COALESCE(account_status, 'undefined') as account_status,
                COALESCE(is_verified, FALSE) as is_verified,
                COUNT(*) as count
            FROM users 
            GROUP BY account_status, is_verified
            ORDER BY account_status, is_verified
        `);

        if (userStats.length > 0) {
            console.log('👥 Usuários por status:');
            userStats.forEach(stat => {
                const verifiedText = stat.is_verified ? '✅ Verificado' : '❌ Não verificado';
                console.log(`   - ${stat.account_status}: ${stat.count} usuários (${verifiedText})`);
            });
        }

        // Registros de verificação pendentes
        const pendingVerifications = await this.executeQuery(`
            SELECT COUNT(*) as count 
            FROM email_verifications 
            WHERE verified = FALSE
        `);

        if (pendingVerifications.length > 0) {
            console.log(`📧 Verificações pendentes: ${pendingVerifications[0].count}`);
        }

        // Últimas tentativas de verificação
        const recentAttempts = await this.executeQuery(`
            SELECT user_id, email, created_at 
            FROM email_verifications 
            ORDER BY created_at DESC 
            LIMIT 5
        `);

        if (recentAttempts.length > 0) {
            console.log('📋 Últimas tentativas de verificação:');
            recentAttempts.forEach(attempt => {
                console.log(`   - Usuário ${attempt.user_id}: ${attempt.email} em ${attempt.created_at}`);
            });
        }

        console.log('✅ Relatório concluído!');
    }

    async runAllFixes() {
        console.log('🚀 INICIANDO RESOLUÇÃO DE PROBLEMAS - VIBE');
        console.log('='.repeat(60));
        console.log(`⏰ Início: ${new Date().toISOString()}`);

        if (!await this.connectDatabase()) {
            return false;
        }

        try {
            await this.fixCooldownIssues();
            await this.fixUserVerificationSystem();
            await this.fixEmailServiceConfig();
            await this.createSummaryReport();

            console.log('\n' + '='.repeat(60));
            console.log('🎉 TODAS AS CORREÇÕES APLICADAS COM SUCESSO!');
            console.log('='.repeat(60));
            console.log('\n✅ Próximos passos:');
            console.log('1. Reiniciar o serviço de email: cd email-service && npm run dev');
            console.log('2. Reiniciar o backend principal');
            console.log('3. Testar login com conta não verificada (deve bloquear)');
            console.log('4. Testar verificação de email (cooldown de 60s)');
            console.log('\n💡 Agora o sistema:');
            console.log('- ✅ Bloqueia login sem verificação de email');
            console.log('- ✅ Usa cooldown de 60 segundos');
            console.log('- ✅ Gerencia status da conta corretamente');
            console.log('- ✅ Remove registros problemáticos automaticamente');

            return true;

        } catch (error) {
            console.error(`\n❌ Erro durante a execução: ${error.message}`);
            return false;
        } finally {
            if (this.connection) {
                await this.connection.end();
                console.log('\n🔌 Conexão com banco fechada');
            }
        }
    }
}

async function main() {
    const resolver = new VibeResolver();
    const success = await resolver.runAllFixes();

    if (success) {
        console.log('\n🎯 Execução concluída com sucesso!');
        process.exit(0);
    } else {
        console.log('\n💥 Execução falhou. Verifique os logs acima.');
        process.exit(1);
    }
}

// Executar se este arquivo for chamado diretamente
main().catch(error => {
    console.error('❌ Erro fatal:', error);
    process.exit(1);
});

export default VibeResolver;
